extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/BloomFilterMergeAggregatorFactory.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.query.aggregation.bloom;
+
+import org.apache.druid.java.util.common.ISE;
+import org.apache.druid.query.aggregation.Aggregator;
+import org.apache.druid.query.aggregation.AggregatorFactory;
+import org.apache.druid.query.aggregation.AggregatorUtil;
+import org.apache.druid.query.aggregation.BufferAggregator;
+import org.apache.druid.query.cache.CacheKeyBuilder;
+import org.apache.druid.query.filter.BloomKFilter;
+import org.apache.druid.segment.ColumnSelectorFactory;
+import org.apache.druid.segment.ColumnValueSelector;
+import org.apache.druid.segment.NilColumnValueSelector;
+
+import java.nio.ByteBuffer;
+import java.util.Collections;
+import java.util.List;
+
+public class BloomFilterMergeAggregatorFactory extends BloomFilterAggregatorFactory
+{
+  private final String fieldName;
+
+  BloomFilterMergeAggregatorFactory(String name, String field, Integer maxNumEntries)
+  {
+    super(name, null, maxNumEntries);
+    this.fieldName = field;
+  }
+
+  @Override
+  public Aggregator factorize(final ColumnSelectorFactory metricFactory)
+  {
+    final ColumnValueSelector<BloomKFilter> selector = metricFactory.makeColumnValueSelector(fieldName);
+    if (selector instanceof NilColumnValueSelector) {
+      throw new ISE("WTF?! Unexpected NilColumnValueSelector");
```

Comment:
[To expand on this comment](https://github.com/apache/incubator-druid/pull/6397#discussion_r247308520), when examining how this would be called my conclusion was that this would be an unexpected condition in the merge aggregator, because the output of a null column from the aggregator would still be a bloom filter with the null or default value which is what this would see. Did I misinterpret or miss a situation where this could actually happen?

Reply:
I just noticed that this is the merge aggregator. Then it probably makes sense, but there should be a comment and the error message should be more descriptive.

But then it appears that `BloomFilterAggregatorFactory` doesn't special-case `NilColumnValueSelector`, however it probably should for performance.

Reply:
After refactor this now special cases `NilColumnValueSelector` and I think behaves in a slightly more correct manner of producing a totally empty bloom filter instead of a bloom filter with a null value

Reply:
Please add a comment